apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -265,80 +254,85 @@ const DanceAiModal: React.FunctionComponent<DanceAiModalProps> = ({
       [FieldKey.BACKGROUND_EFFECT]: backgroundLabels,
       [FieldKey.BACKGROUND_PALETTE]: paletteLabels,
     };
-  };
-
-  const getAllItems = () => {
-    return inputLibraryJson.items.map(item => {
-      return {
-        ...item,
-        available: !inputs.includes(item.id),
-      };
-    });
-  };
-
-  const getItem = (id: string) =>
-    inputLibraryJson.items.find(item => item.id === id);
+  }, []);
 
-  const handleItemClick = (id: string, available: boolean) => {
-    if (available) {
-      // Add item inputs.
-      if (currentInputSlot < SLOT_COUNT) {
-        setInputs([...inputs, id]);
-        setCurrentInputSlot(currentInputSlot + 1);
-        setInputAddCount(inputAddCount + 1);
+  const handleItemClick = useCallback(
+    (id: string, available: boolean) => {
+      setInputs(inputs => {
+        // if we've clicked on an emoji that's not available, remove it from our inputs
+        if (!available) {
+          playSound('ai-deselect-emoji');
+          return inputs.filter(input => input !== id);
+          // if we have more input slots, we can add the emoji
+        } else if (inputs.length < SLOT_COUNT) {
+          playSound('ai-select-emoji');
+          setInputAddCount(inputAddCount => inputAddCount + 1);
+          return [...inputs, id];
+          // otherwise, it's not a remove and we don't have space - do nothing.
+        } else {
+          return inputs;
+        }
+      });
+    },
+    [playSound]
+  );
 
-        playSound('ai-select-emoji');
-      }
-    } else {
-      // Remove item from inputs.
-      setInputs(inputs.filter(input => input !== id));
-      setCurrentInputSlot(currentInputSlot - 1);
+  const startAi = useCallback(() => {
+    generatedEffects.current = {
+      badEffects: getRangeArray(0, BAD_GENERATED_RESULTS_COUNT - 1).map(() =>
+        chooseEffects(inputs, ChooseEffectsQuality.BAD)
+      ),
+      goodEffect: chooseEffects(inputs, ChooseEffectsQuality.GOOD),
+    };
 
-      playSound('ai-deselect-emoji');
-    }
-  };
+    minMaxAssociations.current = calculateMinMax(inputs);
+  }, [calculateMinMax, inputs]);
 
-  const handleGenerateClick = () => {
+  const handleGenerateClick = useCallback(() => {
     startAi();
 
     analyticsReporter.sendEvent(EVENTS.DANCE_PARTY_AI_BACKGROUND_GENERATED, {
       emojis: inputs,
     });
 
     setMode(Mode.GENERATING);
-  };
+  }, [startAi, inputs]);
+
+  const onClose = useCallback(() => dispatch(closeAiModal()), [dispatch]);
 
-  const handleStartOverClick = () => {
+  const handleStartOverClick = useCallback(() => {
     analyticsReporter.sendEvent(EVENTS.DANCE_PARTY_AI_BACKGROUND_RESTARTED, {
       emojis: inputs,
     });
+
     setInputs([]);
-    setCurrentInputSlot(0);
     setGeneratingProgress({step: 0, subStep: 0});
     setGeneratedProgress(0);
     setMode(Mode.SELECT_INPUTS);
-  };
+  }, [inputs]);
```

Comment:
These extra handlers here _would_ trigger re-renders every time the inputs change, but fortunately they're only attached to buttons that appear after all inputs have come into place. So we can get away with leaving them as is.

This could actually be dealt with via the same trick of hopping into the `setInputs` mutator as such:

```
setInputs(inputs => 
  analyticsReporter.sendEvent(EVENTS.DANCE_PARTY_AI_BACKGROUND_RESTARTED, {
    emojis: inputs,
  });
  return inputs
)
```

And then no dependency is required and the callback persists forever. But it's unclear why it's doing that, and since the button it's attached to wouldn't exist until `inputs` are set, I left it as is. If I think of a clever way to pull out the analytics reporting from the handler to make this truly survive forever, I will.